src/libYARP_os/src/yarp/os/Log.cpp

```diff
@@ -35,432 +47,809 @@
-#define GREEN   (colored_output ? "\033[01;32m" : "")
-#define YELLOW  (colored_output ? "\033[01;33m" : "")
-#define BLUE    (colored_output ? "\033[01;34m" : "")
-#define MAGENTA (colored_output ? "\033[01;35m" : "")
-#define CYAN    (colored_output ? "\033[01;36m" : "")
-#define WHITE   (colored_output ? "\033[01;37m" : "")
-#define RED_BG  (colored_output ? "\033[01;41m" : "")
-#define CLEAR   (colored_output ? "\033[00m" : "")
-
 #ifdef YARP_HAS_WIN_VT_SUPPORT
 #include <windows.h>
+#endif
 
-namespace {
-bool yarp_logger_vt_colors_enabled = false;
 
-bool yarp_logger_enable_vt_colors()
+class yarp::os::impl::LogPrivate
 {
-    DWORD handleMode = 0;
-    HANDLE hStdout = GetStdHandle(STD_OUTPUT_HANDLE);
-    bool success = false;
+public:
+    LogPrivate(const char* file,
+               const unsigned int line,
+               const char* func,
+               const LogComponent& comp);
+
+    void log(yarp::os::Log::LogType type,
+             const char *msg,
+             va_list args) const;
+
+    static void print_callback(yarp::os::Log::LogType t,
+                               const char* msg,
+                               const char* file,
+                               const unsigned int line,
+                               const char* func,
+                               double systemtime,
+                               double networktime,
+                               const char* comp_name);
+
+    static void forward_callback(yarp::os::Log::LogType t,
+                                 const char* msg,
+                                 const char* file,
+                                 const unsigned int line,
+                                 const char* func,
+                                 double systemtime,
+                                 double networktime,
+                                 const char* comp_name);
+
+    // Calls the right print and forward callbacks
+    static void do_log(yarp::os::Log::LogType type,
+                       const char* msg,
+                       const char* file,
+                       const unsigned int line,
+                       const char* func,
+                       double systemtime,
+                       double networktime,
+                       const LogComponent& comp_name);
+
+    // This is a LogCallback that calls the print callback that is currently
+    // set, even if this is changed later
+    static void default_print_callback(yarp::os::Log::LogType type,
+                                       const char* msg,
+                                       const char* file,
+                                       const unsigned int line,
+                                       const char* func,
+                                       double systemtime,
+                                       double networktime,
+                                       const char* comp_name)
+    {
+        if (auto cb = current_print_callback.load()) {
```

Comment:
[If-with-init-statement](https://en.cppreference.com/w/cpp/language/if) is a C++17 thing, but YARP supports C++14, yet:

https://github.com/robotology/yarp/blob/4d561eefbdda1dae9870e8c19e0b59a698ddbbcf/cmake/YarpSystemCheck.cmake#L27

https://github.com/robotology/yarp/blob/f6310fb0cab38605115f20b7fe83c52fbd8f41a7/src/libYARP_os/src/CMakeLists.txt#L413

Reply:
Are you sure about this? According to the link you posted, "any init-statement must end with a semicolon `;`". From my limited knowledge of the standard, this is a c++11 (since it uses `auto`) "declaration of a single non-array variable with a brace-or-equals initializer"

Also this test works properly on most compilers supporting c++11... https://godbolt.org/z/zVK5yP

If I understand correctly, the main difference is that using the c++17 init statement, the `x` variable in the example can be used also in the else branch, and it is destroyed before returning to the upper scope, in this case the variable is destroyed at the end of the `if` branch, but since there is no `else` branch here, this should not be an issue...

Reply:
You are totally right and I misread that part, my bad...